### PR TITLE
improve test output, support more problem sizes

### PR DESCRIPTION
**Pull request by elliottbinder**
Added better formatted output for the test driver.
Added an option for setting the number of trials to run.
Added better support for various input sizes.

### CODE REVIEW

**Comment by broderickpt** [comment]
Your fork is missing updates from the upstream.  please 'fetch & merge' from upstream and ping me when you're ready.

**Comment by elliottbinder** [comment]
I have fetched and merged with the latest.

File: examples/3DDFT_mpi/test3DDFT_mpi_1D.cpp [hunk 1]
```diff
@@ -41,10 +45,9 @@ int main(int argc, char* argv[]) {
   bool is_embedded = 0 < atoi(argv[5]);
   bool is_forward  = 0 < atoi(argv[6]);
   bool is_complex  = 0 < atoi(argv[7]);
+  int trials       = atoi(argv[8]);
```
Comment (broderickpt) [inline examples/3DDFT_mpi/test3DDFT_mpi_1D.cpp:48]: Please ensure 'trials' >= 1

**Review by broderickpt — CHANGES_REQUESTED** [review]
I ran a bunch (22) of test cases with the output shown below.  14 cases indicate Pass, 7 indicate Fail, and in 1 case the program aborted.  Can you look at the cases/output and advise if some test cases aren't supported or correct those which shouldn't fail.

Tests:

./test3DDFT_mpi_1D 64 64 64 1 0 1 1 10 1
Problem size: 64 x 64 x 64
Batch size  : 1
Complex     : Yes
Embedded    : No
Direction   : Forward
MPI Ranks   : 1
Times       : 
non-supported fixed library transform
non-supported fixed library transform
non-supported fixed library transform
	Trial 0: 0.749821 seconds
non-supported fixed library transform
non-supported fixed library transform
non-supported fixed library transform
	Trial 1: 0.748884 seconds
non-supported fixed library transform
non-supported fixed library transform
non-supported fixed library transform
	Trial 2: 0.757493 seconds
non-supported fixed library transform
non-supported fixed library transform
non-supported fixed library transform
	Trial 3: 0.748765 seconds
non-supported fixed library transform
non-supported fixed library transform
non-supported fixed library transform
	Trial 4: 0.742776 seconds
non-supported fixed library transform
non-supported fixed library transform
non-supported fixed library transform
	Trial 5: 0.742889 seconds
non-supported fixed library transform
non-supported fixed library transform
non-supported fixed library transform
	Trial 6: 0.742446 seconds
non-supported fixed library transform
non-supported fixed library transform
non-supported fixed library transform
	Trial 7: 0.751478 seconds
non-supported fixed library transform
non-supported fixed library transform
non-supported fixed library transform
	Trial 8: 0.744809 seconds
non-supported fixed library transform
non-supported fixed library transform
non-supported fixed library transform
	Trial 9: 0.74207 seconds
Correct     : Yes

./test3DDFT_mpi_1D 64 64 64 4 0 1 1 10 1
Problem size: 64 x 64 x 64
Batch size  : 4
Complex     : Yes
Embedded    : No
Direction   : Forward
MPI Ranks   : 1
Times       : 
	Trial 0: 0.0128058 seconds
	Trial 1: 0.00396398 seconds
	Trial 2: 0.00399678 seconds
	Trial 3: 0.00395632 seconds
	Trial 4: 0.00397661 seconds
	Trial 5: 0.00397716 seconds
	Trial 6: 0.00408024 seconds
	Trial 7: 0.00395731 seconds
	Trial 8: 0.00396788 seconds
	Trial 9: 0.00403247 seconds
Correct     : Yes

./test3DDFT_mpi_1D 64 64 64 16 0 1 1 1 1
Problem size: 64 x 64 x 64
Batch size  : 16
Complex     : Yes
Embedded    : No
Direction   : Forward
MPI Ranks   : 1
Times       : 
	Trial 0: 0.0296404 seconds
Correct     : Yes

./test3DDFT_mpi_1D 64 64 64 16 0 0 1 1 1
Problem size: 64 x 64 x 64
Batch size  : 16
Complex     : Yes
Embedded    : No
Direction   : Inverse
MPI Ranks   : 1
Times       : 
	Trial 0: 0.02825 seconds
Correct     : Yes

./test3DDFT_mpi_1D 64 64 64 4 0 0 1 1 1
Problem size: 64 x 64 x 64
Batch size  : 4
Complex     : Yes
Embedded    : No
Direction   : Inverse
MPI Ranks   : 1
Times       : 
	Trial 0: 0.0118505 seconds
Correct     : Yes

./test3DDFT_mpi_1D 64 64 64 1 0 0 1 1 1
Problem size: 64 x 64 x 64
Batch size  : 1
Complex     : Yes
Embedded    : No
Direction   : Inverse
MPI Ranks   : 1
Times       : 
non-supported fixed library transform
non-supported fixed library transform
non-supported fixed library transform
	Trial 0: 0.755876 seconds
Correct     : Yes

./test3DDFT_mpi_1D 256 256 256 1 0 0 1 1 1
Problem size: 256 x 256 x 256
Batch size  : 1
Complex     : Yes
Embedded    : No
Direction   : Inverse
MPI Ranks   : 1
Times       : 
non-supported fixed library transform
non-supported fixed library transform
non-supported fixed library transform
	Trial 0: 1.07331 seconds
Correct     : Yes

./test3DDFT_mpi_1D 256 256 256 4 0 0 1 1 1
Problem size: 256 x 256 x 256
Batch size  : 4
Complex     : Yes
Embedded    : No
Direction   : Inverse
MPI Ranks   : 1
Times       : 
	Trial 0: 0.339116 seconds
Correct     : Yes

./test3DDFT_mpi_1D 256 256 256 16 0 0 1 1 1
Problem size: 256 x 256 x 256
Batch size  : 16
Complex     : Yes
Embedded    : No
Direction   : Inverse
MPI Ranks   : 1
Times       : 
Memory access fault by GPU node-5 (Agent handle: 0x1078480) on address 0x901e000. Reason: Unknown.
Aborted

./test3DDFT_mpi_1D 256 256 256 16 0 0 1 0 1
Problem size: 256 x 256 x 256
Batch size  : 16
Complex     : Yes
Embedded    : No
Direction   : Inverse
MPI Ranks   : 1
Times       : 
Correct     : No

./test3DDFT_mpi_1D 256 256 256 16 0 0 1 -2 1
Problem size: 256 x 256 x 256
Batch size  : 16
Complex     : Yes
Embedded    : No
Direction   : Inverse
MPI Ranks   : 1
Times       : 
Correct     : No

./test3DDFT_mpi_1D 128 128 128 4 1 1 1 3 1
Problem size: 128 x 128 x 128
Batch size  : 4
Complex     : Yes
Embedded    : Yes
Direction   : Forward
MPI Ranks   : 1
Times       : 
	Trial 0: 0.121914 seconds
	Trial 1: 0.0888339 seconds
	Trial 2: 0.0880541 seconds
Correct     : No

./test3DDFT_mpi_1D 128 128 128 1 1 1 1 3 1
Problem size: 128 x 128 x 128
Batch size  : 1
Complex     : Yes
Embedded    : Yes
Direction   : Forward
MPI Ranks   : 1
Times       : 
non-supported fixed library transform
non-supported fixed library transform
non-supported fixed library transform
	Trial 0: 0.962892 seconds
non-supported fixed library transform
non-supported fixed library transform
non-supported fixed library transform
	Trial 1: 0.952475 seconds
non-supported fixed library transform
non-supported fixed library transform
non-supported fixed library transform
	Trial 2: 0.954052 seconds
Correct     : Yes

./test3DDFT_mpi_1D 128 128 128 4 0 1 1 3 1
Problem size: 128 x 128 x 128
Batch size  : 4
Complex     : Yes
Embedded    : No
Direction   : Forward
MPI Ranks   : 1
Times       : 
	Trial 0: 0.0530071 seconds
	Trial 1: 0.0322 seconds
	Trial 2: 0.0335861 seconds
Correct     : No

./test3DDFT_mpi_1D 64 64 64 4 0 1 1 3 1
Problem size: 64 x 64 x 64
Batch size  : 4
Complex     : Yes
Embedded    : No
Direction   : Forward
MPI Ranks   : 1
Times       : 
	Trial 0: 0.01303 seconds
	Trial 1: 0.00402028 seconds
	Trial 2: 0.00402313 seconds
Correct     : Yes

./test3DDFT_mpi_1D 64 64 64 16 0 1 1 3 1
Problem size: 64 x 64 x 64
Batch size  : 16
Complex     : Yes
Embedded    : No
Direction   : Forward
MPI Ranks   : 1
Times       : 
	Trial 0: 0.0306307 seconds
	Trial 1: 0.0175066 seconds
	Trial 2: 0.0175114 seconds
Correct     : Yes

./test3DDFT_mpi_1D 64 64 64 16 1 1 1 3 1
Problem size: 64 x 64 x 64
Batch size  : 16
Complex     : Yes
Embedded    : Yes
Direction   : Forward
MPI Ranks   : 1
Times       : 
	Trial 0: 0.0631718 seconds
	Trial 1: 0.0515471 seconds
	Trial 2: 0.0511798 seconds
Correct     : Yes

./test3DDFT_mpi_1D 64 64 64 16 0 0 1 3 1
Problem size: 64 x 64 x 64
Batch size  : 16
Complex     : Yes
Embedded    : No
Direction   : Inverse
MPI Ranks   : 1
Times       : 
	Trial 0: 0.028523 seconds
	Trial 1: 0.0137613 seconds
	Trial 2: 0.0138077 seconds
Correct     : Yes

./test3DDFT_mpi_1D 64 64 64 16 0 0 0 3 1
Problem size: 64 x 64 x 64
Batch size  : 16
Complex     : No
Embedded    : No
Direction   : Inverse
MPI Ranks   : 1
Times       : 
	Trial 0: 0.0206984 seconds
	Trial 1: 0.00792473 seconds
	Trial 2: 0.00786575 seconds
Correct     : Yes

./test3DDFT_mpi_1D 64 64 64 16 0 1 0 3 1
Problem size: 64 x 64 x 64
Batch size  : 16
Complex     : No
Embedded    : No
Direction   : Forward
MPI Ranks   : 1
Times       : 
	Trial 0: 0.0224531 seconds
	Trial 1: 0.00973932 seconds
	Trial 2: 0.00980903 seconds
Correct     : No

./test3DDFT_mpi_1D 64 64 64 16 4 1 0 3 1
Problem size: 64 x 64 x 64
Batch size  : 16
Complex     : No
Embedded    : Yes
Direction   : Forward
MPI Ranks   : 1
Times       : 
	Trial 0: 0.0381094 seconds
	Trial 1: 0.0262097 seconds
	Trial 2: 0.0261582 seconds
Correct     : No

./test3DDFT_mpi_1D 64 64 64 16 0 9 0 3 1
Problem size: 64 x 64 x 64
Batch size  : 16
Complex     : No
Embedded    : No
Direction   : Forward
MPI Ranks   : 1
Times       : 
	Trial 0: 0.0213226 seconds
	Trial 1: 0.00931411 seconds
	Trial 2: 0.0111305 seconds
Correct     : No

**Comment by elliottbinder** [comment]
There were some issues with batch != 1, which I believe I have fixed, but I can only get them to work when using either smaller input sizes, batch sizes, or more MPI ranks from what you tested with. My guess is that the issue could be from too large of buffers than supported by library APIs (e.g. MPI uses ints for send count), or running out of memory. I'd have to check on that, but we might want to wrap our memory allocators with error checking code to start.

**Comment by broderickpt** [comment]
How do I change # MPI ranks?  
Now we're down to 3 of the original batch of 22 tests failing:

./test3DDFT_mpi_1D 128 128 128 4 1 1 1 3 1
./test3DDFT_mpi_1D 128 128 128 1 1 1 1 3 1
./test3DDFT_mpi_1D 128 128 128 4 0 1 1 3 1

we have 256^3 tests run OK and 128^3 fail -- surely that's not a case where the input is too large (yes 2 cases are embedded but 1 is not).  Can you determine what is going on?  What sizes are valid and what guidance can be given to a person 'kicking the tires' who won't know your test code?

**Comment by elliottbinder** [comment]
MPI ranks can be changed by passing a different value to the `-n` argument of `mpirun`, as in `mpirun -n 4 ./test3DDFT_mpi_1D ...`.

**Comment by elliottbinder** [comment]
After some investigation, it looks like it might be due to numerical error, which can accumulate as the size of the signal increases. I'll make the appropriate changes to the test routine.

**Pull request merged**